GLMakie/src/GLAbstraction/GLTypes.jl

```diff
@@ -397,14 +402,30 @@ function RenderObject(
     end
 
     buffers = filter(((key, value),) -> isa(value, GLBuffer) || key === :indices, data)
-    program = gl_convert(to_value(program), data) # "compile" lazyshader
+    program = gl_convert(context, to_value(program), data) # "compile" lazyshader
     vertexarray = GLVertexArray(Dict(buffers), program)
+    require_context(context)
+
+    # Validate context of things in RenderObject
+    if GLMAKIE_DEBUG[]
+        require_context(program.context, context)
+        require_context(vertexarray.context, context)
+        for v in values(data)
+            if v isa TextureBuffer
+                require_context(v.buffer.context, context)
+                require_context(v.texture.context, context)
+            elseif v isa GPUArray
+                require_context(v.context, context)
+            end
+        end
+    end
 
     # remove all uniforms not occurring in shader
     # ssao, instances transparency are special for rendering passes. TODO do this more cleanly
     special = Set([:ssao, :transparency, :instances, :fxaa, :num_clip_planes])
     for k in setdiff(keys(data), keys(program.nametype))
         if !(k in special)
+            !haskey(buffers, k) && (data[k] isa GPUArray) && free(data[k])
```

Comment:
This disallows optional shared resources. E.g. if we were to pass the same OpenGL texture to two renderobjects and one does not include a uniform for it, it would get deleted here.